### ISSUE TITLE
Proposal: Add more info to SelectionModelSelectionChangedArgs

### ISSUE THREAD

**Issue by grokys** [microsoft/microsoft-ui-xaml#2247]
# Proposal: Add more info to SelectionModelSelectionChangedArgs

## Summary

As described in #1645 `SelectionModelSelectionChangedEventArgs` carries no information about what changed in a selection. This is a proposal to add that information, based on modifications done to a port of `SelectionModel` in Avalonia. An implementation of this (in C#) can be found at https://github.com/AvaloniaUI/Avalonia/pull/3470.

## Rationale

* When a selection changes, it is important to be able to tell what changed, otherwise once has to store a secondary copy of the current selection and diff if with the new selection on change
* Secondary: it can be useful to be able to batch changes

## Scope

| Capability | Priority |
| :---------- | :------- |
| Include information on items/indexes added/removed to/from selection | Must |
| Support batching of changes | Could |

## Important Notes

The proposed API for `SelectionModelSelectionChangedEventArgs` is the following:

```csharp
public class SelectionModelSelectionChangedEventArgs : EventArgs
{
	/// <summary>
	/// Gets the indices of the items that were removed from the selection.
	/// </summary>
	public IReadOnlyList<IndexPath> DeselectedIndices { get; }

	/// <summary>
	/// Gets the indices of the items that were added to the selection.
	/// </summary>
	public IReadOnlyList<IndexPath> SelectedIndices { get; }

	/// <summary>
	/// Gets the items that were removed from the selection.
	/// </summary>
	public IReadOnlyList<object?> DeselectedItems { get; }

	/// <summary>
	/// Gets the items that were added to the selection.
	/// </summary>
	public IReadOnlyList<object?> SelectedItems { get; }
}
```

The class contains selected and deselected collections containing both indexes and items. These collections are lazy (using a modified version of the existing `SelectedItems` collection) and populated in the following scenarios:

- `SelectedItems`/`DeselectedItems` are always populated when items are added/removed to/from the selection
- In addition `DeselectedItems` is populated when selected items are removed from the `ItemSource`
- `SelectedIndices`/`DeselectedIndices` are always populated when items are added/removed to/from the selection
- `DeselectedIndices` is _not_ populated when selected items are removed from the `ItemSource` because these items no longer have an index

In addition, you may want to consider batching changes to the selection. This is done with the following method on `SelectionModel`:

```
public IDisposable Update();
```

Calling this method increases the operation count, and disposing the returned object decreases the operation count. When the operation count reaches 0 a `SelectionChanged` event is raised.

Rather than going into more detail in this proposal, it may be useful instead to link to the unit tests for the feature in Avalonia. They start at the following line:

https://github.com/AvaloniaUI/Avalonia/blob/3910c7c856ffe18e6ac5fb08cf81c55d46c0c14a/tests/Avalonia.Controls.UnitTests/SelectionModelTests.cs#L892

**Comment by chingucoding** [comment]
A few questions:

1. Is there a benefit in keeping the IndexPaths for the changed items? It seems like we would just store more information that is rarely used but will impact performance negatively.
2. How do the DeselectedIndices map to the DeselectedItems? Do they map at all or could it be that the first entry in DeselectedIndices wouldn't actually be the index of the first DeselectedItem?
3. What exactly do you mean by "lazily" evaluated? Since every item in the ItemsSource is pass by reference I think, those collection should/could probably be empty lists when created and not related to the actual list.
4. Having this "batch" processing would create a huge overhead (collect changes, what if we undo an operation, check when we raise batch etc.). Is there really a benefit to having this? I can't think of a use case right now, where it makes sense to wait for let's say 4 selection changes and then only raise the event. If you don't want to process it immediately, you can either save them yourself and deal with it later or you are actually only interested in the selections at some point later (e.g. a filepicker) and you don't need that event at all.
5. If we would add the batch processing, we should think of a more natural way of increasing/decreasing the counter. Having an IDisposable and calling it or disposing it to increase/decrease really seems unnatural to me and a bit complicated.

**Comment by grokys** [comment]
Hi @chingucoding, good questions!

> Is there a benefit in keeping the IndexPaths for the changed items? It seems like we would just store more information that is rarely used but will impact performance negatively.

Yes, if you have duplicate items in your item source then you'll need the indexes to identify them. This can come up when for example, displaying a list of `string`s. Here's an issue that was opened against Avalonia when someone hit upon this: https://github.com/AvaloniaUI/Avalonia/issues/2004

> How do the DeselectedIndices map to the DeselectedItems? Do they map at all or could it be that the first entry in DeselectedIndices wouldn't actually be the index of the first DeselectedItem?

When both collections are present, then the ordering will be the same. They're both iterating over the same collection of `IndexRange`s.

> What exactly do you mean by "lazily" evaluated? Since every item in the ItemsSource is pass by reference I think, those collection should/could probably be empty lists when created and not related to the actual list.

When I mean is that the collections aren't materialized as lists which contain every item. Like `SelectionModel.SelectedIndices` and `SelectionModel.SelectedItems`, they're implemented as a view over a collection of `IndexRange`s. See the existing [`SelectedItems` class](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/SelectedItems.h) for how it works currently for `SelectedItems`.

> Having this "batch" processing would create a huge overhead (collect changes, what if we undo an operation, check when we raise batch etc.). Is there really a benefit to having this? I can't think of a use case right now, where it makes sense to wait for let's say 4 selection changes and then only raise the event. If you don't want to process it immediately, you can either save them yourself and deal with it later or you are actually only interested in the selections at some point later (e.g. a filepicker) and you don't need that event at all.

The state for batch processing is actually required internally anyway (at least in my implementation) because many operations internally consist of more than one physical operation on the selection. See for example the [`SelectedIndex` setter](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/SelectionModel.cpp#L136-L137) which first clears the selection and then selects the new item. Because these operations are actually taking place across any number of `SelectionNode`s, there needs to be a central place to record the change.

> If we would add the batch processing, we should think of a more natural way of increasing/decreasing the counter. Having an IDisposable and calling it or disposing it to increase/decrease really seems unnatural to me and a bit complicated.

An `IDisposable` makes sense in C# because it means you'll usually surround the operation with a `using` block. I guess C++ tends to use RAII so might make sense to expose it in a different way to C++?

**Comment by chingucoding** [comment]
> Yes, if you have duplicate items in your item source then you'll need the indexes to identify them. This can come up when for example, displaying a list of strings. Here's an issue that was opened against Avalonia when someone hit upon this: AvaloniaUI/Avalonia#2004

I think we are boxing non pass by reference, but I am not entirely sure on that. And for pass by reference, is it really a problem? After all you would have to add the EXACT same item twice, which most likely doesn't make a whole lot sense when talking about selection.

> When both collections are present, then the ordering will be the same. They're both iterating over the same collection of IndexRanges.

Maybe we should provide a more robust way for this then or get rid of one of the sides. One possible return value could be a tuple of <IndexPath,Object> so those mappings don't get "destroyed" so easily.

> So if you have a collection with 1,000,000 items, and do a "select all", the change is stored as a an IndexRange of [0-999999]. SelectedItems and SelectedIndices are views over that IndexRange rather than being collections holding 1,000,000 items each.

Thanks for the explanation. With that in mind, maybe we should just return the selected indices and let the developer get the items? Otherwise we would have to spent a lot of resources on just creating the event args, which would definitely be bad.

> The state for batch processing is actually required internally anyway (at least in my implementation) because many operations internally consist of more than one physical operation on the selection. See for example the SelectedIndex setter which first clears the selection and then selects the new item. Because these operations are actually taking place across any number of SelectionNodes, there needs to be a central place to orchestrate the change, which is essentially what is being exposed here.

Yes that's a bit problematic. Internally, we just pass parameters to indicate whether to raise the event or not.

> See #1872 for an example of where batch processing is useful.

I think in that case there are other workarounds that are better, for example you could calculate the diff of the range and deselect the appropriate range. That's of course more difficult but also in my opinion more cleaner than suppressing the event for some operations and then raise it again.

> An IDisposable makes sense in C# because it means you'll usually surround the operation with a using block. I guess C++ tends to use RAII so might make sense to expose it in a different way to C++?

Semantically an IDisposable indicates that an object can release unmanaged ressources. If I understand you correctly, how you would try to solve the batch problem is definitely not about clearing ressources but about raising a counter, maybe something like a semaphore.

It just feels counterintuitive to call `Dispose` to indicate the selectionmodel that it can raise it's event now (or it should not), especially if we are behaving more like a counter.

Regarding the interop between C++ and C#, we are using the winrt projections. Since C# is one of our main audiences but not our only one, we should try to create a solution that feels intuitive for everyone, and not just for C#.

Also what do you mean by "[...] you'll usually surround the operation with a using block" with regard to an IDisposable? (Real question, is this some pattern I haven't heard of before?)

**Comment by ranjeshj** [comment]
@grokys Thank you for creating the proposal. 

I believe [ICloseable](https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.IClosable) in WinRT maps to IDispose in C# so we can use that to get the same behavior.

```
DeselectedIndices is not populated when selected items are removed from the ItemSource because these items no longer have an index
```
I'm curious. why not pass the indices before it got removed ? 

```
Because the collections in SelectionModelSelectionChangedEventArgs are lazily evaluated, if the user makes a change to the items source in a SelectionChanged handler, then reading these collections could result in bad information. Does this case need to be handled? e.g.
```

If changing selection in the SelectionChanged handler is not supported (reentrancy) we should really guard against that because that can lead to subtle bugs. I think this might be supported currently because the event is raised at the very end. If this is specific to when reading the new properties, we can guard against that specifically.

**Comment by ranjeshj** [comment]
@grokys, adding a sample of how to use the batching above would be useful. I think batching can be quite helpful to avoid raising a bunch of intermediate selectionchanged events.

**Comment by chingucoding** [comment]
Couldn't we run into the risk of locking the SelectionChanged event if we have that "counter" behavior?

**Comment by ranjeshj** [comment]
Let me try to explain what I understand of the batching behavior and we can figure out the correct way to achieve it. Assume that I want to do a bunch of selection operations but all of them come one after the other. Currently each individual change would cause a SelectionChanged to fire and everyone listening to the event would react to it. Instead I want to do all the updates and then raise just one event that captures all the changes.

@grokys has achieved this through this pattern.

```
 using (selectionModel.Update())
            {
                selectionModel.Deselect(1);
                selectionModel.Select(4);
            }
```

Update is a bit confusing. Some more options that provide to get the same behavior.

```
selectionModel.StartBatch();
selectionModel.Deselect(1);
selectionModel.Select(4);
selectionModel.CommitBatch()

auto batch = selectionModel.CreateBatch();
selectionModel.Deselect(1);
selectionModel.Select(4);
batch.Commit();

using(selectionModel.CreateBatch())
{
  selectionModel.Deselect(1);
  selectionModel.Select(4);
}

```

In all these cases, what we want is SelectionChanged to be raised just once at the end.

@MikeHillberg Is there a preference/precedence for this? I think comp might have something similar.

**Comment by grokys** [comment]
> I think we are boxing non pass by reference, but I am not entirely sure on that. And for pass by reference, is it really a problem? After all you would have to add the EXACT same item twice, which most likely doesn't make a whole lot sense when talking about selection.

The most common case for this is displaying a list of strings: take a look at the issue I linked. If you display the list: `[ "Foo", "Foo", "Bar" ]` and then say "Foo" has been selected, which "Foo" are we talking about?

> I think in that case there are other workarounds that are better, for example you could calculate the diff of the range and deselect the appropriate range. That's of course more difficult but also in my opinion more cleaner than suppressing the event for some operations and then raise it again.

Seems like you'll be forcing a user to raise multiple selection changed events and making the class harder to use, all to avoid exposing some internal state that's there anyway?

> Semantically an IDisposable indicates that an object can release unmanaged ressources. If I understand you correctly, how you would try to solve the batch problem is definitely not about clearing ressources but about raising a counter, maybe something like a semaphore.

That may have been the original meaning of `IDisposable` but in my experience it's often used as a RAII pattern for C#. Some prior art:

- https://reactiveui.net/docs/handbook/obsolete/collections/reactive-list#suppressing-notifications
- http://introtorx.com/Content/v1.0.10621.0/03_LifetimeManagement.html#IDisposable

I'm sure there's more - I'm pretty sure there's examples of this in the BCL, though I can't remember where...

> maybe something like a semaphore.

A semaphore as traditionally defined is for synchronization across threads, what do you mean here?

> Also what do you mean by "[...] you'll usually surround the operation with a using block" with regard to an IDisposable? (Real question, is this some pattern I haven't heard of before?)

I mean you could do something like this:

```csharp
using (selection.Update())
{
    selection.ClearSelection();
    selection.SelectRange(...);
}
```

Or with a C# 8 [using declaration](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#using-declaration):

```csharp
{
    using operation = selection.Update();
    selection.ClearSelection();
    selection.SelectRange(...);
}
```

> Couldn't we run into the risk of locking the SelectionChanged event if we have that "counter" behavior?

Of course, if the `Update()` wasn't placed in a `using` block or in a using statement scope. Not sure how one would implement this without running that risk however.

**Comment by chingucoding** [comment]
How would the using pattern work in C++? Can we just "transfer" that or wouldn't it be used easily?

The "Start" and "Commit" batch seems a bit more intuitive, but also could result in race conditions with multi threading.

**Comment by grokys** [comment]
> The naming 'Update' is a bit confusing in my opinion. Some more options that provide to get the same behavior. 
>  Is there a preference/precedence for this? I think comp might have something similar.

Yeah, if there's existing terminology for this, would be best to use that.

> The "Start" and "Commit" batch seems a bit more intuitive, but also could result in race conditions with multi threading.

I don't think `SelectionModel` is designed to be thread-safe anyway?

**Comment by grokys** [comment]
> I'm curious. why not pass the indices before it got removed ?

I figured this would be confusing if sometimes the indexes referred to indexes in the current collection, and sometimes to indexes in a previous state of the collection. Also what would you do with those indexes anyway? They're no longer valid, so I can't see any way to make use of them.

> If changing selection in the SelectionChanged handler is not supported (reentrancy)

I think reentrancy is useful here - for example if you want to always have an item selected, you would do this by listening for a selection changed event and re-selecting an item if nothing is selected.

**Comment by chingucoding** [comment]
> The most common case for this is displaying a list of strings: take a look at the issue I linked. If you display the list: [ "Foo", "Foo", "Bar" ] and then say "Foo" has been selected, which "Foo" are we talking about?

The "Foo" you are whose pointer you have. At least in ItemsSourceView, those get boxed so you can identify them even if you have the value "Foo" twice.

> Seems like you'll be forcing a user to raise multiple selection changed events and making the class harder to use, all to avoid exposing some internal state that's there anyway?

It's not really an internal state, we don't really track "should we raise now" or not as some sort of state, operations just request to not raise it, but yes I see your point. In the issue linked, you could probably only raise the event once, but in some other scenarios it might not be possible. Maybe it would be better to have.

> That may have been the original meaning of IDisposable but in my experience it's often used as a RAII pattern for C#. Some prior art:
> https://reactiveui.net/docs/handbook/obsolete/collections/reactive-list#suppressing-notifications
> http://introtorx.com/Content/v1.0.10621.0/03_LifetimeManagement.html#IDisposable
> I'm sure there's more - I'm pretty sure there's examples of this in the BCL, though I can't remember where...

Well they use the "using" pattern, but we should not just pick an interface that was designed for something different.

> A semaphore as traditionally defined is for synchronization across threads, what do you mean here?

The fact that a semaphore uses an counter internally but also defines the appropriate actions. Update and Dispose are not actions you would expect to use to increase/decrease a counter.

> Of course, if the Update() wasn't placed in a using block or in a using statement scope. Not sure how one would implement this without running that risk however.

Guess that people just have to live with that risk then.

> I don't think SelectionModel is designed to be thread-safe anyway?

I am not sure, I think we didn't take any precautions with that. However since with event handlers, you can not be entirely sure when code does get executed, so you could potentially have a problem here. Maybe this is irrelevant though.

> I think reentrancy is useful here - for example if you want to always have an item selected, you would do this by listening for a selection changed event and re-selecting an item if nothing is selected.

Agree, we should be able to change selection inside the handler (though you have to watch out to not create a stackoverflow currently right?).

---

One alternative to the batch idea could be to just expose functions that suppress that event, just like we do internally. That way we wouldn't have to think about locks and possible problems of scoping.

Your thoughts @grokys and @ranjeshj ?

**Comment by ranjeshj** [comment]
> One alternative to the batch idea could be to just expose functions that suppress that event, just like we do internally. That way we wouldn't have to think about locks and possible problems of scoping.
> 
> Your thoughts @grokys and @ranjeshj ?

It is not just suppressing the event though right ? when the event finally gets raised, I still want to know **everything** that has changed.

**Comment by chingucoding** [comment]
Oh, that's true. We could have an internal batch then on which we add the changes. And the next time the event get's raised we will just show the changes of all previously suppressed "unhandled" changes and then clear the lists.

**Comment by grokys** [comment]
> The "Foo" you are whose pointer you have. At least in ItemsSourceView, those get boxed so you can identify them even if you have the value "Foo" twice.

This might be a difference between C# and C++, in C# there's no easy way to do this I don't think. Also strings can be interned.

> Well they use the "using" pattern, but we should not just pick an interface that was designed for something different.

Perhaps, I'm not aware of any guidance that states that `IDisposable` should _only_ be used for managed resources, but I could have missed it ;)

>> One alternative to the batch idea could be to just expose functions that suppress that event, just like we do internally. That way we wouldn't have to think about locks and possible problems of scoping.
> It is not just suppressing the event though right ? when the event finally gets raised, I still want to know everything that has changed.

Exactly, if you have items `[0-5]` selected then you clear the selection and reselect 5, the event will be raised with `[0-4]` as the deselected items.

**Comment by chingucoding** [comment]
> Perhaps, I'm not aware of any guidance that states that IDisposable should only be used for managed resources, but I could have missed it ;)

Well it says "Provides a mechanism for releasing unmanaged resources." in the documentation, and I don't think that is what we are doing here.

> This might be a difference between C# and C++, in C# there's no easy way to do this I don't think. Also strings can be interned.

If we cross the "border of ABI" C# strings will get boxed into winrt::hstring if I recall correctly, which support that. Interning of course would be a problem here, though it is a bit "weird" to have interned strings and add them multiple times to your selectionmodel. But we should probably support it?

**Comment by grokys** [comment]
@chingucoding, one piece of information that might be useful for you regarding needing indexes, is that in Avalonia, `SelectionModel` will be used provide the selection state for e.g. `ListBox`.

`ListBox` in UWP/WPF current handles duplicate items and so Avalonia needs to support this by extension.

If you have a list of `string` and receive a notification that an item `"Foo"` has been selected, you need to find the `ListBoxItem` that holds that item. The quickest way to do this is by looking it up by index, otherwise you need to enumerate every single string in your `ListBox` to find the correct one (even if you are able to distinguish between duplicates somehow).

You mention elsewhere:

> With that in mind, maybe we should just return the selected indices and let the developer get the items?

Two responses to this:

- When items are removed from the source collection, they no longer have an index
- `SelectionModel` already exposes `SelectedItems` as well as `SelectedIndices`, which I think is a recognition that both ways of looking at the selection can be useful

**Comment by chingucoding** [comment]
> @chingucoding, one piece of information that might be useful for you regarding needing indexes, is that in Avalonia, SelectionModel will be used provide the selection state for e.g. ListBox.

> ListBox in UWP/WPF currently handles duplicate items and so Avalonia needs to support this by extension.

> If you have a list of string and receive a notification that an item "Foo" has been selected, you need to find the ListBoxItem that holds that item. The quickest way to do this is by looking it up by index, otherwise you need to enumerate every single string in your ListBox to find the correct one 
> (even if you are able to distinguish between duplicates somehow).

Yes, that's true. One could use IndexOf to get the index. Though if we already have that internally we should expose it. Agreed, Thanks for the explanation!

> Two responses to this:
> * When items are removed from the source collection, they no longer have an index

Should we remove the index too in that case? And how would your mapping work if we have an item in the collection, but it has no valid index? Wouldn't the collections get out of sync with batch processing?

> * SelectionModel already exposes SelectedItems as well as SelectedIndices, which I think is a recognition that both ways of looking at the selection can be useful

That's a good point. 

> My personal feeling is that that documentation is outdated. As I say, I could be wrong here though ;)

I think there's a difference between how some people use it and what it was designed to be used for. Maybe you should create an issue or PR for [the doc page](https://github.com/dotnet/dotnet-api-docs/blob/master/xml/System/IDisposable.xml) :)

**Comment by anawishnoff** [comment]
Getting on this thread late, but loving the very detailed and technical discussion going on here.

One initial question I had while reading the proposal:
> All of the collections are unpopulated when a change to the item source results in just the selected indexes being adjusted

If an item is added to the ItemsSource, is it guaranteed that it won't change the indices of the other items in the list? If so, shouldnt Selected/DeselectedIndices collections change when an item is added to the ItemsSource, regardless of if it's selected? The indices of selected items would be changing.

Although a lot of the details about the batching implementation have admittedly  gone over my head a bit, I think it is worthwhile to have as a part of the selection model as I see the importance of wanting to only raise one SelectionChanged event rather than one for each change.

**Comment by grokys** [comment]
> If an item is added to the ItemsSource, is it guaranteed that it won't change the indices of the other items in the list?

No, it's quite likely that it _will_ change the indices of the other items in the list. Let's first clarify what's happening here:

- We have a list of chars as the `ItemsSource`: `['a', 'b', 'c', 'd']` 
- We have index 2 selected, i.e. `'c'`
- Now we insert a new item at index 0 into the `ItemsSource`: `itemSource.Insert(0, 'x)`
- Now our selection hasn't changed, but the index of the selected item has
- `SelectionChanged` needs to be fired here, what with?

My decision was to fire `SelectionChanged` with all of the collections empty wasn't easy, but I debated it (with myself) and came to the following conclusions:

- The selection hasn't actually _changed_ - the same items are selected
- Although the index of the selected item **has** changed - because `ItemsSource` has now been mutated the previous index now has no meaning: there's no way to use the old index to do anything useful with the current state of `ItemsSource`
- Suppose we do want to populate `DeselectedIndices` with the old index. In all other cases `DeselectedIndices` refers to an index in the current `ItemsSource` - how do we tell the difference between "The deselected index refers to the current `ItemsSource`" and "The deselected index refers to a state of `ItemsSource` that is now lost"?

Because I couldn't come up with a use-case for the old index, I decided to signal this case by just leaving everything empty.

I definitely don't believe that this is the optimal solution however, and if we can come up with a use-case where the old index is useful then it may be that we need to find a way to signal this. Note that I did mention this subject in https://github.com/microsoft/microsoft-ui-xaml/issues/1645#issuecomment-576689668:

> Should there be a flag for "indexes changed" where the selection hasn't changed, but the indexes of some of the selection has changed?

Maybe we do need such a flag, but I figured "empty collections all round" was as good a flag as any ;)

**Comment by ranjeshj** [comment]
Is there value in having a separate IndicesChanged list ? Also, how did you handle the reset case ? I think it would be useful to clarify what the args will hold for each scenario below. I think I covered everything, please update the list if I missed something.

1. Item(s) selected
2. Items(s) deselected
3. [Collection changes](https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized.notifycollectionchangedaction?view=netframework-4.8)
     (a) Items added, some selected items may have changed indices now
     (b) Items moved, some selected items may have changed indices now
     (c) Items removed, some selected items might have been removed
     (d) Items replaced, some selected items might be replaced (same index, but new item)
     (e) Reset - 
              i. source supports IKeyIndexMapping, in which case we can recreate what changed and make this a combination of a,b,c,d
              ii. No IKeyIndexMapping, we are truly blind as to what changed, so just notify somehow that a reset happened.
 4. What happens in cases above when there is no bound ItemsSource.

**Comment by chingucoding** [comment]
Maybe, instead of raising SelectionChanged, we could introduce a new event "IndicesChanged" or something along the lines. For that event, we could then provide a event args appropriate for that case. For example a mapping between the old and new index of the selected items.

The problem with handling this in SelectionChanged is what would happen if there are more and more corner cases? We would then have to bloat the SelectionEventArgs with properties that may be empty/false in most cases.

**Comment by ranjeshj** [comment]
Yeah.. That crossed my mind as well. I think it would make sense to look at how this would be used from a control perspective. What would the handler code look like to handle all the cases.

**Comment by grokys** [comment]
Perhaps. Again I think we'd need some sort of use-case though.

In the case of Avalonia all controls that use `SelectionModel` also listen to the `INotifyCollectionChanged.CollectionChanged` event on the items source because they also need to update the containers that _aren't_ selected. That means that additional information about how the indexes have changed in `SelectionModel` isn't necessary.

I suspect that this would be the case with most controls that want to use `SelectionModel`, but again, hard to know without more use-cases.

**Comment by chingucoding** [comment]
Fair point. Since without an IObservableCollection we aren't able to detect indices changes correctly, maybe we shouldn't inform about index changes then?

After all, if we are not giving information about what changed, there isn't much to do anyway except essentially "reload" everything.

I think the most common (and also designed) use case is using it with an ItemsRepeater. And mapping items to their selected state there is most commonly done by referring to the item in the collection since the ItemsRepeater only accepts simple indices but no IndexPath. And maybe there you wouldn't gain much by getting information "object a is now at index 3, not 4" since you can also access your UIObjects using the object.

So I am inclined to say,, this wouldn't help much at all. Maybe we shouldn't raise SelectionChanged in that case?

**Comment by grokys** [comment]
> So I am inclined to say, this wouldn't help much at all. Maybe we shouldn't raise SelectionChanged in that case?

Possibly! Though note that `SelectionChanged` _is_ currently raised in this case (which is why I left it in).

**Comment by chingucoding** [comment]
That's a good point indeed. Currently however the event is mostly useless in regard to clean up work on unselected items and similiar situations, so every event cause behaves the same.

I think in the NavigationView case which uses the SelectionModel, we are handling collection changes ourselves and wouldn't benefit from knowing the difference or being notified by the SelectionModel that the list has changed. But maybe this is something @ojhad can comment on, as he was the one who implemented the switch from ListView to ItemsRepeater (and SelectionModel) with #1818 .

**Comment by anawishnoff** [comment]
After reading the last few comments it seems like there isn't a super common use case in which you'd need access to the changed indices after selection, and I see the dilemma that SelectionChanged is called in this instance, as @grokys laid out in [this example](https://github.com/microsoft/microsoft-ui-xaml/issues/2247#issuecomment-610907742) above.

However I also see that it's important that `SelectionChanged` be called if the indices of selected items change - in case you have two duplicate strings and one becomes unselected, just knowing the list of objects that are selected won't inform you which of the duplicate strings is currently selected. Or is there a solution for that? If so, it seems like we don't need a Deselected/SelectedIndices list or an IndicesChanged list/event. Correct me if I'm wrong!  It just seems like having the object itself should be enough.

**Comment by chingucoding** [comment]
I think there isn't a solution for having the same object multiple times except using indices.

In most cases, having just the object is enough, but in some corner cases that can get problematic. To cover every case, we should probably include the indices too. 

Would it make sense to decouple the event args from batching and have two proposals? Adding the event args is something we should probably do sooner than the batching, since the lack of any information what changed is a huge issue if we talk about performance or ease of use.

**Comment by ranjeshj** [comment]
Sounds like two orthogonal problems to solve, so yes, they can be two separate proposals if that helps stage them independently. Thanks!

**Comment by chingucoding** [comment]
@grokys Do you want to create a separate proposal for batching or would you like me to do that?

**Comment by grokys** [comment]
Feel free to do it! I've not got a whole lot of time at the moment.

**Comment by chingucoding** [comment]
@grokys Created #2330 for this. Could you update your proposal to only include the SelectionModelSelectionChangedArgs ? :) Thank you!

**Comment by anawishnoff** [comment]
Thanks for splitting up the issues! What are the loose ends here? 

For the IndicesChanged list, providing that after an item is removed would only cause confusion and not be useful - for that reason, if it's possible we should return an IndicesChanged list for other collection change events, but not for removal.

**Comment by chingucoding** [comment]
I am not entirely sure when exactly we should raise the event:
- When items get inserted?
- When items get removed?

On the one hand, developers in some instances need to know that indices changed, on the other hand raising the SelectionChanged event with empty lists could be confusing.

I think the rest of this proposal is pretty clear: We have a list of deselected and selected indices and items where the n-th item of the selected indices list is the index of the n-th item in the selected items list. I would prefer having an object that contains both informations in a single object, instead of having two lists that are loosely coupled, but both solutions seem fine to me.

**Comment by anawishnoff** [comment]
> I am not entirely sure when exactly we should raise the event:

Are you referring to the SelectionChanged event?

> 
> * When items get inserted?
> * When items get removed?
> 
> On the one hand, developers in some instances need to know that indices changed, on the other hand raising the SelectionChanged event with empty lists could be confusing.

When items are inserted, DeselectedIndices will return a list of the indices that were selected before this item's entrance (if it caused items to move around and the selection to change). If items are removed, and one of the removed items was previously selected, DeselectedIndices will return the index of the previously selected item that is now removed - that number is pretty useless and points to something else now. I think it would be appropriate in that case to just return null for DeselectedIndices to avoid confusion, rather than an empty list.

However, I still think that SelectionChanged should be raised whenever the selected indices or selected items change, whether that be by actual selection/deselection or by collection changes. Instead of controlling when to fire the event, it may make more sense to control which types of selection changes provide which types of information. @chingucoding and @ranjeshj let me know your thoughts.

> I think the rest of this proposal is pretty clear: We have a list of deselected and selected indices and items where the n-th item of the selected indices list is the index of the n-th item in the selected items list. I would prefer having an object that contains both informations in a single object, instead of having two lists that are loosely coupled, but both solutions seem fine to me.

What kind of object are you thinking of here? A new type that has SelectedIndices and SelectedItems as list-type properties or just a list of two lists?

**Comment by chingucoding** [comment]
> Are you referring to the SelectionChanged event?

Yes, sorry about the confusion.

>  I think it would be appropriate in that case to just return null for DeselectedIndices to avoid confusion, rather than an empty list.

That's true, returning null is probably be better. Though I am not such a fan of returning null in events as this would require additional code handling (check for null, then go through lists).

>  Instead of controlling when to fire the event, it may make more sense to control which types of selection changes provide which types of information. 

If we are really technical, the items that have been selected to not change when we insert items, only the index of those items (which we can't inform the user about what the new one is). And since we don't tell the user what exactly is the new selection index wise, maybe a separate event is more logical here? If we were able to provide some sort of list of the old indices and the corresponding new ones, it might make more sense to raise SelectionChanged.

> What kind of object are you thinking of here? A new type that has SelectedIndices and SelectedItems as list-type properties or just a list of two lists?

I was more thinking about an object that contains an IndexPath and an object, so we have to lists of those objects and you can iterate over those and ask for the index or object (or both) in a single iteration. E.g:

```c#
public SelectionTuple
{
    IndexPath Index;
    Object Item;
}
```

**Comment by anawishnoff** [comment]
Hmm, okay, I'm definitely starting to come around to the idea of having a separate event for IndicesChanged. It makes more sense from an implementation side, but I'm not sure if it's something that would be so intuitive from the developer's side - i.e. developers may not think to listen for that event as they would selection changed. 

However, if we had a separate event we could still return SelectedIndices with the SelectionChanged event as it might be useful. Looking for some clarity here: For IndicesChanged, it would fire every time an index gets matched up with a new item. Would we return a list of SelectionTuples as you drew out above? Or would it be a list of now-selected indices and now deselected indices? How could you use those two lists to compare to previous selections?

I'm also concerned that providing a more vague event like IndicesChanged makes it difficult to determine the best IndicesChangedEventArgs - it's very possible people would be using this for more than just selection.

**Comment by chingucoding** [comment]
> It makes more sense from an implementation side, but I'm not sure if it's something that would be so intuitive from the developer's side - i.e. developers may not think to listen for that event as they would selection changed.

Agreed, that is one potential downside. The change if indices is somewhat a selectionchange and somewhat it is not, so it is hard to argue here what is the right event to handle that kind of change.


> However, if we had a separate event we could still return SelectedIndices with the SelectionChanged event as it might be useful

Totally agree, I think we should still return the indices with the items that have been selected.

> For IndicesChanged, it would fire every time an index gets matched up with a new item. Would we return a list of SelectionTuples as you drew out above? Or would it be a list of now-selected indices and now deselected indices? How could you use those two lists to compare to previous selections?

I think the best case would be something like "Old index -> New index" mapping that developers than can process as they wish. That way I think we have maximum transparency on what has happened. Because that is essentially what happens, some item moved from index  `a` to index `b`, and if you need the index, you would probably know that your selected item sits at index `a` which you could then update.

> I'm also concerned that providing a more vague event like IndicesChanged makes it difficult to determine the best IndicesChangedEventArgs - it's very possible people would be using this for more than just selection.

Maybe we could name it like "SelectedItemsIndicesChanged" so it is clear that we are only caring about items that have been selected and whose index has changed. Having a "SelectedItemsIndicesChangedEventArgs" object would also probably prevent it from becoming to general as the name already is quite specific compared to IndiciesChangedEventArgs.

**Comment by anawishnoff** [comment]
Agreed - something specific like SelectedItemsIndicesChanged or even just SelectionChangedIndices would get the message across. This might be the easiest way to control what info we pass back when theres a selection or collection change. @ranjeshj What do you think of this method of having two separate events, so we can control where we provide information on indices (preventing us from providing misleading or confusion info on every SelectionChanged event)?

Also, returning an object where the old index maps to the new index sounds like the right idea. Perhaps a slight change to the SelectionTuple object where it provides 
```csharp
IndexPath OldIndex;
IndexPath NewIndex;
object Item;
```

**Comment by chingucoding** [comment]
> Also, returning an object where the old index maps to the new index sounds like the right idea. Perhaps a slight change to the SelectionTuple object where it provides
> ```c# 
> IndexPath OldIndex;
> IndexPath NewIndex;
> object Item;
> ```

Great idea! I think for the indices changed event that is all the information one might need to deal with that event.
